prov/gni/src/gnix_tags.c

```diff
@@ -0,0 +1,564 @@
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "rdma/fabric.h"
+#include "rdma/fi_tagged.h"
+
+#include "gnix_tags.h"
+
+#include "gnix.h"
+#include "gnix_util.h"
+
+struct gnix_tag_storage_ops list_ops;
+struct gnix_tag_storage_ops hlist_ops;
+struct gnix_tag_storage_ops kdtree_ops;
+
+struct gnix_tag_search_element {
+	uint64_t tag;
+	uint64_t ignore;
+	void *context;
+	uint64_t flags;
+	int use_src_addr_matching;
+	struct gnix_address *addr;
+};
+
+/**
+ * @brief converts gnix_tag_list_element to gnix_fab_req
+ *
+ * @param elem  slist element embedded in a gnix_fab_req
+ * @return pointer to gnix_fab_req
+ */
+static inline struct gnix_fab_req *__to_gnix_fab_req(
+		struct gnix_tag_list_element *elem)
+{
+	struct gnix_fab_req *req;
+
+	req = container_of(elem, struct gnix_fab_req, tle);
+
+	return req;
+}
+
+/**
+ * @brief determines if a req matches the address parameters
+ *
+ * @param addr_to_find    address to find in tag storage
+ * @param addr            stored address
+ * @return 0 if the request does not match the parameters, 1 otherwise
+ */
+static inline int __req_matches_addr_params(
+		struct gnix_address *addr_to_find,
+		struct gnix_address *addr)
+{
+	return (GNIX_ADDR_UNSPEC(*addr) ||
+			GNIX_ADDR_EQUAL(*addr_to_find, *addr));
+}
+
+int _gnix_req_matches_params(
+		struct gnix_fab_req *req,
+		uint64_t tag,
+		uint64_t ignore,
+		uint64_t flags,
+		void *context,
+		int use_src_addr_matching,
+		struct gnix_address *addr,
+		int matching_posted)
+{
+	int valid_request;
+
+	/* adding some error checking to the first condition
+	 *
+	 * if the context is null, then FI_PEEK | FI_CLAIM should fail
+	 */
+	if ((flags & FI_CLAIM) && (flags & FI_PEEK))
+		valid_request = (context != NULL &&
+				context != req->tle.context);
+	else if ((flags & FI_CLAIM) && !(flags & FI_PEEK))
+		valid_request = (req->tle.context != NULL &&
+				context == req->tle.context);
+	else
+		valid_request = req->tle.context == NULL;
+
```

Comment:
Not sure if this is something a compiler can figure out these days, but it seems like you should just return here if `valid_request == 0`.